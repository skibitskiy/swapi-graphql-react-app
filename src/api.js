const baseUrl = 'https://swapi.dev/api/';

const fetchData = async (type, page = 1) => {
    const url = new URL(type, baseUrl);
    url.searchParams.append('page', page);
    const response = await fetch(url).then((res) => res.json());
    return response
}

const fetchResources = async (type, page = 1) => {
    const { results: resources } = await fetchData(type, page);
    return resources
}

const fetchTitles = async (type, page) => {
    const resources = await fetchResources(type, page);
    return resources.map((resource) => {
        const { name, title, url } = resource;
        return {
            name: name || title,
            url
        }
    });
}

const fetchPageCount = async (type) => {
    const { count } = await fetchData(type);
    return Math.ceil((parseInt(count) || 0) / 10);
}

const fetchResource = async (url) => {
    const response = await fetch(url).then((res) => res.json());
    return response
}

export {
    fetchData,
    fetchResources,
    fetchTitles,
    fetchPageCount,
    fetchResource
}
