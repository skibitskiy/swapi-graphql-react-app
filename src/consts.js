const FILMS = 'films';
const PEOPLE = 'people';
const PLANETS = 'planets';
const SPECIES = 'species';
const STARSHIPS = 'starships';
const VEHICLES = 'vehicles';
const CHOOSE_OBJECT = 'Выбери объект';

const selectItems = {
    [FILMS]: 'Фильмы',
    [PEOPLE]: 'Персонажи',
    [PLANETS]: 'Планеты',
    [SPECIES]: 'Расы',
    [STARSHIPS]: 'Звездолеты',
    [VEHICLES]: 'Транспорт'
};

export {
    selectItems,
    FILMS,
    PEOPLE,
    PLANETS,
    SPECIES,
    STARSHIPS,
    VEHICLES,
    CHOOSE_OBJECT
}