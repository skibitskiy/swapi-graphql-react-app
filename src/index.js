import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './components/App.js';
import * as serviceWorker from './serviceWorker';
import { ApolloClient, InMemoryCache, ApolloProvider  } from '@apollo/client';

const serverURI = process.env.SERVER;
const client = new ApolloClient({
  uri: serverURI || 'http://localhost:1166/graphql',
  cache: new InMemoryCache({
    possibleTypes: {
      IResource: ['Film', 'People', 'Species', 'Vehicle', 'Starship', 'Planet']
    }
  })
});

ReactDOM.render(
  <ApolloProvider client={client}>
    <React.StrictMode>
      <App />
    </React.StrictMode>
  </ApolloProvider>,
  document.getElementById('root')
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();
