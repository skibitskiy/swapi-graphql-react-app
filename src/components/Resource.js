import React, { useCallback, useState, useMemo, memo } from 'react';
import { useLazyQuery } from '@apollo/client';
import { Accordion, AccordionSummary, AccordionDetails, Typography } from '@material-ui/core';
import { ExpandMore } from '@material-ui/icons';

import Progress from './Progress';

import { getResource } from '../api';

const getPrettyDescription = (resource) =>
    Object.entries(resource)
        .map((res) => <>{ res.join(': ').slice(0, 20) }<br/></>)

function Resource(props) {
    const { name, url, resourceType } = props;
    const id = useMemo(() => parseInt(url.match(/(\d+)\/$/)[1]), [url]);

    const [fetchResource, { loading, data: resource }] = useLazyQuery(getResource, { variables: { id, resourceType } });

    const [expanded, setExpanded] = useState(false);
    const onExpand = useCallback(() => {
        setExpanded((state) => !state)
        if (expanded === false && resource === undefined) {
            fetchResource()
        }
    }, [expanded, resource, fetchResource])
    
    return (
        <Accordion expanded={expanded} onChange={onExpand}>
            <AccordionSummary expandIcon={<ExpandMore />}>
                <Typography>{ name }</Typography>
            </AccordionSummary>
            <AccordionDetails>
                { 
                    loading || resource === undefined
                        ? <Progress/>
                        : getPrettyDescription(resource.getResource)
                }
            </AccordionDetails>
        </Accordion>
    )
}

export default memo(Resource);
