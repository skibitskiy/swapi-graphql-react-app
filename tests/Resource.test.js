import '@testing-library/jest-dom/extend-expect';
import waitForExpect from 'wait-for-expect';
import React from 'react';
import * as test from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import user from '@testing-library/user-event';

import Resource from '../src/components/Resource';
import { getResource } from '../src/api';
import { apolloCache } from './helpers';

describe('Resource component', () => {
    it('Snapshot', async () => {
        const props = {
            name: 'Luke Skywalker',
            resourceType: 'people',
            url: 'http://swapi.dev/api/people/1/'
        };
        const mocks = [
            {
                request: {
                    query: getResource,
                    variables: {
                        resourceType: props.resourceType,
                        id: parseInt(props.url.match(/(\d+)\/$/)[1])
                    }
                },
                result: {
                    data: {
                        getResource: {
                            url: props.url,
                            __typename: 'People'
                        }
                    }
                }
            }
        ];
        const { container } = test.render(
            <MockedProvider mocks={mocks} cache={apolloCache()}>
                <Resource {...props}/>
            </MockedProvider>
        );

        // start state
        expect(container).toMatchSnapshot();

        // loading state
        test.act(() => {
            const resource = test.screen.getByText(props.name);
            user.click(resource);
        });
        expect(container).toMatchSnapshot();

        // final state
        await waitForExpect(() => {
            expect(test.screen.getByText('url', { exact: false })).toBeInTheDocument();
        });
        expect(container).toMatchSnapshot();
    })
})
