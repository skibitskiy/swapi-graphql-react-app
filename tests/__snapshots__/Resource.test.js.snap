// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource component Snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Luke Skywalker
        </p>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
              >
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <div
                    class="MuiCircularProgress-root makeStyles-progress-1 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                    role="progressbar"
                    style="width: 40px; height: 40px;"
                  >
                    <svg
                      class="MuiCircularProgress-svg"
                      viewBox="22 22 44 44"
                    >
                      <circle
                        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                        cx="44"
                        cy="44"
                        fill="none"
                        r="20.2"
                        stroke-width="3.6"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Resource component Snapshot 2`] = `
<div>
  <div
    class="MuiPaper-root MuiAccordion-root Mui-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="true"
      class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content Mui-expanded"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Luke Skywalker
        </p>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon Mui-expanded MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container"
      style="min-height: 0px; height: 0px; transition-duration: 0ms;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
              >
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <div
                    class="MuiCircularProgress-root makeStyles-progress-1 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                    role="progressbar"
                    style="width: 40px; height: 40px;"
                  >
                    <svg
                      class="MuiCircularProgress-svg"
                      viewBox="22 22 44 44"
                    >
                      <circle
                        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                        cx="44"
                        cy="44"
                        fill="none"
                        r="20.2"
                        stroke-width="3.6"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Resource component Snapshot 3`] = `
<div>
  <div
    class="MuiPaper-root MuiAccordion-root Mui-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="true"
      class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content Mui-expanded"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Luke Skywalker
        </p>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon Mui-expanded MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container MuiCollapse-entered"
      style="min-height: 0px; height: auto; transition-duration: 0ms;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
            >
              __typename: People
              <br />
              url: http://swapi.de
              <br />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
